---
title: "U method intall and run"
author: "Yaniv Stein"
date: "2025-02-02"
output: html_document
---

# Umethod

**Umethod** is an R package designed for identifying unique markers in single-cell data sets, Use the FindUniqueMarkers function on a Seurat object after clustering to get the most unique markers for your clusters. CreateImageData function can be used with U markers (or any other markers) for each cluster for downstream analysis, visualizing the markers on Visium HD spatial data.

## ðŸš€ Installation

To install Umethod from GitHub:

#### Install remotes if you haven't already
install.packages("remotes")

#### Install Umethod from GitHub
remotes::install_github("YanuvS-Dev/Umethod")

#### Load the package
library(Umethod)

## FindUniqueMarkers algorithm

```{r, echo=FALSE, out.width='50%'}
knitr::include_graphics("images/UmethodImage.png")
```


### ðŸ“ˆ Example Usage

**Reference:** [Lee, Hae-Ock, et al. Nature genetics (2020)]. *"Lineage-dependent gene expression programs influence the immune landscape of colorectal cancer."*, available from ArrayExpress E-MTAB-8410.

To load the reananlyzed example data set directly from GitHub, use the following code in R

```{r,warning = FALSE, message = FALSE}
# Load Umethod
library(Umethod)
library("scCustomize")
library(cowplot)
library(ggplot2)
library(svMisc)

# Load the published data set (replace with the actual data loading code)
seurat_Full <- readRDS(system.file("extdata", "ColonSinglecellDataLeesUmethodSubsampled10.rds", package = "Umethod"))

# Apply Umethod functions, if there are small/mixed clusters, their name should be added to smallcluster variable to omit them.
# The progress bar prints weird massages, in this rmd file I suppress it
genes_list <- FindUniqueMarkers(
    obj = seurat_Full,
    group_by = "Celltype",
    method = "none",
    smallcluster = c("CAFelse", "SmallElse"),
    progresstext = F)

# gene_list is the marker list ordered by score and cluster
head(genes_list)

# Choose thresholds 
Uscore <- 0.25
p_in <- 0.4
p_out <- 0.3

# Pulling the top 5 markers and the name of the top U marker for each cluster

genesetshort <- unlist(sapply(split(genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in & genes_list$P_out < p_out,],genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in& genes_list$P_out < p_out,]$Cluster),function(x){x[[1]][1]}))

genesetlong <- unique(unlist(sapply(split(genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in& genes_list$P_out < p_out,],genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in& genes_list$P_out < p_out,]$Cluster),function(x){x[[1]][1:5]})))

genesetlong

#Order the cluster that had at least one marker that passed threshold, as you want them to be plotted.
clusterorder<- c("CAF","General.Fibro","Adamdec1.Fibro","Normal.Muscle","Sox6..Stroma","CAP.else","Endothelial","Macrofague","T.cells","B.cells","Plasma","Epithelial","Cancer")
# Ordering the clusters that had any umarkers from genesetlong for dotplot
indclusters <- rep(NA,dim(genesetlong)[2])
for(i in 1:dim(genesetlong)[2]){indclusters[i] <- which(colnames(genesetlong) == clusterorder[i])}

genesetlong <- genesetlong[,indclusters]
genesetshort <- genesetshort[indclusters]

```

# Plotting the results of the top U markers for each cluster

```{r,fig.width=27, fig.height=14, out.width='100%', dpi=300,warning = FALSE, message = FALSE, echo=FALSE}
plot_grid(plot_grid(DimPlot(object = seurat_Full, reduction = "UMAP_on_harmony",pt.size = 0.5, group.by = "Celltype"),
                    DotPlot(seurat_Full,features = c(genesetlong),group.by = "Celltype",scale = F) + theme(axis.text.x = element_text(angle = 90,vjust = -0.0001)),ncol = 1),
          FeaturePlot(object = seurat_Full, features = genesetshort, cols = c("gray", "blue"),reduction = "UMAP_on_harmony",ncol = 3,order = T),ncol = 2)

```

