---
title: "U-method Installation and tutorial"
author: "Yaniv Stein"
date: "2025-04-09"
output: html_document
---

# U-method: Identify Unique Markers in Single-Cell Data  

**U-method** is an R package designed to identify **unique markers** in single-cell datasets. It provides methods for finding the most distinct markers for each cluster in a **Seurat object** and visualizing them in **Visium HD spatial data**.

## Installation  

To install **U-method** from GitHub:

```r
# Install devtools if you haven't already
install.packages("devtools")

# Install Umethod package from GitHub
devtools::install_github("YanuvS-Dev/Umethod")

# Load the package
library(Umethod)
```

## FindUniqueMarkers Algorithm  

The `FindUniqueMarkers` function identifies the most **unique markers** for each cluster in a Seurat object.

```{r, echo=FALSE}
knitr::include_graphics("images/UmethodImage.png")
```

## Example Usage  

### Reference Dataset  

This example uses a **reanalyzed** dataset from:

- **Lee, Hae-Ock, et al. Nature Genetics (2020)**  
- *"Lineage-dependent gene expression programs influence the immune landscape of colorectal cancer."*  
- Available: [ArrayExpress E-MTAB-8410](https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-8410/)  

### Load Example Data 

```{r,warning = FALSE, message = FALSE}
# Load the U-method package
library(Umethod)
library(cowplot)
library(ggplot2)
library(svMisc)
library(Seurat)
library(reshape2)

# Load the published data set (replace with the actual data loading code)
rds_url <- "https://github.com/YanuvS-Dev/Umethod/raw/master/inst/extdata/ColonSinglecellDataLeesUmethodSubsampled10.rds"
seurat_Full <- readRDS(url(rds_url, "rb"))

# Apply U-method functions, if there are small/mixed clusters, their name should be added to omitCluster variable to omit them from calculation.
genes_list <- FindUniqueMarkers(
    obj = seurat_Full,
    group_by = "Celltype",
    method = "BH",
    omitCluster = c("CAFelse", "SmallElse"))

# gene_list is the marker list, ordered by score and cluster
head(genes_list)

# Choose thresholds 
Uscore <- 0.25
p_in <- 0.4
```

### Selecting Top Markers  

```{r,warning = FALSE, message = FALSE}
genesetshort <- unlist(sapply(
    split(genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in, ], 
          genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in, ]$Cluster), 
    function(x) { x[[1]][1] }
))

genesetlong <- unique(unlist(sapply(
    split(genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in, ], 
          genes_list[genes_list$Uscore > Uscore & genes_list$P_in > p_in, ]$Cluster), 
    function(x) { x[[1]][1:5] }
)))

genesetlong
```

### Plotting Unique Markers  

```{r,warning = FALSE, message = FALSE}

clusterorder<- c("CAF","General Fibro","Adamdec1 Fibro","Normal Muscle","Sox6+ Stroma","CAP else","Endothelial","Macrofague","T-cells","B cells","Plasma","Epithelial","Cancer")
# Ordering the clusters that had any umarkers from genesetlong for dotplot
indclusters <- rep(NA,dim(genesetlong)[2])
for(i in 1:dim(genesetlong)[2]){indclusters[i] <- which(colnames(genesetlong) == clusterorder[i])}

genesetlong <- genesetlong[,indclusters]
genesetshort <- genesetshort[indclusters]

```

# Plotting the results of the top U markers for each cluster

```{r,fig.width=27, fig.height=14, out.width='100%', dpi=300,warning = FALSE, message = FALSE, echo=FALSE}
plot_grid(plot_grid(DimPlot(object = seurat_Full, reduction = "UMAP_on_harmony",pt.size = 0.5, group.by = "Celltype"),
                    DotPlot(seurat_Full,features = c(genesetlong),group.by = "Celltype",scale = F) + theme(axis.text.x = element_text(angle = 90,vjust = -0.0001)),ncol = 1),
          FeaturePlot(object = seurat_Full, features = genesetshort, cols = c("gray", "blue"),reduction = "UMAP_on_harmony",ncol = 3,order = T),ncol = 2)

```

## Visualizing Markers on Visium HD  

### Load Visium HD Data  

To begin, only two essential files are required for analysis. Example Visium HD data can be accessed from:

Oliveira, Michelli F., et al. (2024). Characterization of immune cell populations in the tumor microenvironment of colorectal cancer using high-definition spatial profiling. bioRxiv. [DOI: 2024-06]

In this example, we focus on one Normal sample and one Tumor sample. However, we strongly encourage users to apply the U-Method to all five available samples to gain deeper insights into spatial marker distribution and expression patterns.

#### Normal adjacent sample - NAT5

```{r,warning = FALSE, message = FALSE}
library(arrow)
colon.data <- Read10X("C:/myGithub/Uemethod_Bigfiles/VisiumHDcolon/NAT5/8um/filtered_feature_bc_matrix/")

poaraq <- read_parquet("C:/myGithub/Uemethod_Bigfiles/VisiumHDcolon/NAT5/8um/spatial/tissue_positions.parquet")


markers <- c(genesetlong)[!is.na(genesetlong)]
markers <- ifelse(markers == "ECSCR.1","ECSCR",markers)
markers <- markers[markers %in% row.names(colon.data)]

seurat_object <- CreateImageData(bc_matrix = colon.data,poaraq = poaraq,markers = markers)
```

### Calculate Signature Scores  

```{r,warning = FALSE, message = FALSE}
# Plotting the signature expression of the top 5 markers or less if there aren't 5 
SignaturebyCell <- as.data.frame(matrix(NA,ncol = dim(genesetlong)[2],nrow = nrow(seurat_object@meta.data)))
for(i in 1:dim(genesetlong)[2])
{
  x <- genesetlong[,i]
  mdf <- seurat_object@meta.data[,colnames(seurat_object@meta.data) %in% x]
  if(is.null(dim(mdf)))
  {
    SignaturebyCell[,i] <- mdf
    print(paste("Only used",colnames(seurat_object@meta.data)[colnames(seurat_object@meta.data) %in% x],"For",colnames(genesetlong)[i],sep = " "))
  }else{
    SignaturebyCell[,i] <- apply(mdf,1,mean,na.rm = T) 
  }
  print(paste("Finished average expression calculation of",colnames(genesetlong)[i]))
}

names(SignaturebyCell) <- colnames(genesetlong)
colnames(SignaturebyCell) <- paste("Signature",colnames(SignaturebyCell),sep = ".")

seurat_object <- AddMetaData(seurat_object,cbind(seurat_object@meta.data,SignaturebyCell))
```

### Plot Spatial Expression  

Using ggplot to plot the results of the top 5 u-marker average expression

## Prepare Data

```{r,warning = FALSE, message = FALSE}
beforemelt <- seurat_object@meta.data[,colnames(seurat_object@meta.data) %in% c("pxl_row_in_fullres","pxl_col_in_fullres",colnames(SignaturebyCell))]
aftermelt <- melt(beforemelt, id.vars = c("pxl_row_in_fullres", "pxl_col_in_fullres"))
names(aftermelt)[3] <- "Class"
aftermelt$value <- ifelse(aftermelt$value == 0, NA, aftermelt$value)
```

## Generate Plots

```{r,warning = FALSE, message = FALSE}
g <- list()
color_palette <- c("brown1", "turquoise1", "chartreuse", "magenta", "lightblue", "darkviolet", "#619CFF", "#7F7F7F", "darkolivegreen1", "orange", "ghostwhite", "#C4B239", "darkorange1")

for (i in colnames(SignaturebyCell)) {
  index <- which(colnames(SignaturebyCell) == i)
  g[[index]] <- ggplot(aftermelt[aftermelt$Class == i & !is.na(aftermelt$value), ], 
                        aes(x = pxl_col_in_fullres, y = pxl_row_in_fullres, color = Class, alpha = value/max(value))) +
    geom_point(size = 0.01) +
    theme_void() +
    scale_y_reverse() +
    guides(colour = guide_legend(override.aes = list(size = 10))) +
    labs(title = paste("Signature:", i)) +
    scale_color_manual(values = color_palette[index]) +
    theme(
      plot.background = element_rect(fill = "black"),
      legend.position = "right",
      legend.justification = "center",
      legend.box.margin = margin(0, 10, 0, 10),
      plot.margin = margin(10, 20, 10, 10),
      legend.text = element_text(color = "white", face = "bold", family = "sans", size = 20, angle = 90)
    ) +
    scale_alpha(guide = 'none')
}
```

## Display Plots - Patient 5 normal sample

```{r,fig.width=27, fig.height=14, out.width='100%', dpi=300,warning = FALSE, message = FALSE, echo=FALSE}
plot_grid(g[[12]],g[[13]],g[[1]],g[[2]],g[[3]],g[[4]],g[[5]],g[[6]],g[[7]],g[[9]],g[[10]],g[[11]],ncol = 3)
```




```{r,warning = FALSE, message = FALSE,include=F}

colon.data <- Read10X("C:/myGithub/Uemethod_Bigfiles/VisiumHDcolon/CRC5/8um/filtered_feature_bc_matrix/")

poaraq <- read_parquet("C:/myGithub/Uemethod_Bigfiles/VisiumHDcolon/CRC5/8um/spatial/tissue_positions.parquet")


markers <- c(genesetlong)[!is.na(genesetlong)]
markers <- ifelse(markers == "ECSCR.1","ECSCR",markers)
markers <- markers[markers %in% row.names(colon.data)]

seurat_object <- CreateImageData(bc_matrix = colon.data,poaraq = poaraq,markers = markers)
```

```{r,warning = FALSE, message = FALSE,include=F}
# Plotting the signature expression of the top 5 markers or less if there arent 5 
SignaturebyCell <- as.data.frame(matrix(NA,ncol = dim(genesetlong)[2],nrow = nrow(seurat_object@meta.data)))
for(i in 1:dim(genesetlong)[2])
{
  x <- genesetlong[,i]
  mdf <- seurat_object@meta.data[,colnames(seurat_object@meta.data) %in% x]
  if(is.null(dim(mdf)))
  {
    SignaturebyCell[,i] <- mdf
    print(paste("Only used",colnames(seurat_object@meta.data)[colnames(seurat_object@meta.data) %in% x],"For",colnames(genesetlong)[i],sep = " "))
  }else{
    SignaturebyCell[,i] <- apply(mdf,1,mean,na.rm = T) 
  }
  print(paste("Finished average expression calculation of",colnames(genesetlong)[i]))
}

names(SignaturebyCell) <- colnames(genesetlong)
colnames(SignaturebyCell) <- paste("Signature",colnames(SignaturebyCell),sep = ".")

seurat_object <- AddMetaData(seurat_object,cbind(seurat_object@meta.data,SignaturebyCell))
```


```{r,warning = FALSE, message = FALSE,include=F}
beforemelt <- seurat_object@meta.data[,colnames(seurat_object@meta.data) %in% c("pxl_row_in_fullres","pxl_col_in_fullres",colnames(SignaturebyCell))]
aftermelt <- melt(beforemelt, id.vars = c("pxl_row_in_fullres", "pxl_col_in_fullres"))
names(aftermelt)[3] <- "Class"
aftermelt$value <- ifelse(aftermelt$value == 0, NA, aftermelt$value)
```


```{r,warning = FALSE, message = FALSE,include=F}
g <- list()
color_palette <- c("brown1", "turquoise1", "chartreuse", "magenta", "lightblue", "darkviolet", "#619CFF", "#7F7F7F", "darkolivegreen1", "orange", "ghostwhite", "#C4B239", "darkorange1")

for (i in colnames(SignaturebyCell)) {
  index <- which(colnames(SignaturebyCell) == i)
  g[[index]] <- ggplot(aftermelt[aftermelt$Class == i & !is.na(aftermelt$value), ], 
                        aes(x = pxl_col_in_fullres, y = pxl_row_in_fullres, color = Class, alpha = value/max(value))) +
    geom_point(size = 0.01) +
    theme_void() +
    scale_y_reverse() +
    guides(colour = guide_legend(override.aes = list(size = 10))) +
    labs(title = paste("Signature:", i)) +
    scale_color_manual(values = color_palette[index]) +
    theme(
      plot.background = element_rect(fill = "black"),
      legend.position = "right",
      legend.justification = "center",
      legend.box.margin = margin(0, 10, 0, 10),
      plot.margin = margin(10, 20, 10, 10),
      legend.text = element_text(color = "white", face = "bold", family = "sans", size = 20, angle = 90)
    ) +
    scale_alpha(guide = 'none')
}
```

## Display Plots - Patient 5 tumor sample

```{r,fig.width=27, fig.height=14, out.width='100%', dpi=300,warning = FALSE, message = FALSE, echo=FALSE}
plot_grid(g[[12]],g[[13]],g[[1]],g[[2]],g[[3]],g[[4]],g[[5]],g[[6]],g[[7]],g[[9]],g[[10]],g[[11]],ncol = 3)
```

## Notes on Signature Expression figure

* **Normal Samples:** CAF and CAPelse signatures are not expressed.
* **Tumor Samples:** Cancer signature expression differs from Epithelial signature expression only in Tumor samples. 




ðŸš€ Now you're ready to detect unique single-cell markers using the U-method!


